### ISSUE TITLE
Repeated applications of `arrayUpdate` are very slow

### ISSUE THREAD

**Issue by travitch** [GaloisInc/crucible#373]
Setting a large number of concrete values in arrays using the `arrayUpdate` function is very slow.  I'm not entirely sure what the cause is.  I had a use case where iteratively applying `arrayUpdate` took about 20 minutes, while using `arrayUpdateAtIdxLits` was instant.

**Comment by robdockins** [comment]
It's probably this, from `abstractEval`

```
    ArrayMap _ bRepr m d ->
      withAbstractable bRepr $
        Map.foldl' (\av e -> avJoin bRepr (f e) av) (f d) (Hash.hashedMap m)
```

Similar to issues we're encountering with semiring sums, repeated building of these terms causes quadratic work as we compute abstract domain information and hashes, and also the bitvector abstract domain calculations are currently quite expensive (CF #353)

**Comment by robdockins** [comment]
I've pushed a number of changes to the branch `rwd/annotated-boolmap`.  This includes changes to the way the `ArrayMap` syntax former is implemented.  I'm intereted to know if this addresses the performance issues you're seeing.

**Comment by robdockins** [comment]
I'm pretty sure that the performance problem identified here was fixed via #387.  Please reopen if this is still a problem.
